### DISCUSSION TITLE
ruleset rule for no new line before first element when multi-line?

### DISCUSSION THREAD

**Discussion by GlennBags** [squizlabs/PHP_CodeSniffer#3415] [Q&A]
Hi,

Saw that there's a means to do add rules in the ruleset.xml.

I have a weird thing happening where the beautifier is adding a new line in an `if` statement if the conditional is multi-line, but not otherwise:
![image](https://user-images.githubusercontent.com/1598183/130663551-c26e7bd4-18cb-4dcc-955e-7b134b6e7522.png)

Is there a rule that I can add to keep it from adding the new lines?

Thanks,
Glenn

EDIT: Hmm, just read in PSR-12:
"Expressions in parentheses MAY be split across multiple lines, where each subsequent line is indented at least once. When doing so, the first condition MUST be on the next line".
Apparently that's where it's coming from. 
However, would still be curious how to suppress that in the case we decide on that as an exception.

**Comment by jrfnl** [comment]
@GlennBags Yes, PSR-12 has specific requirements, so that sniff may not be the right fit for you.

You may want to try out the `PEAR.ControlStructures.MultiLineCondition` sniff, which I think will more closely match your needs.

**Reply by jrfnl** [reply]
Alternatively, you could try to exclude just that particular error code from PSR-12. I think it's the `PSR12.ControlStructures.ControlStructureSpacing.FirstExpressionLine` error code.

You can exclude it like so:
```xml
<rule ref="PSR12">
    <exclude name="PSR12.ControlStructures.ControlStructureSpacing.FirstExpressionLine"/>
</rule>
```

**Reply by GlennBags** [reply]
Awesome, thanks.  Will give it a try tomorrow.

**Reply by GlennBags** [reply]
BTW, is there a manual/listing where all of these sniffs are listed? I was unable to find, other than the pear stuff, but even that was kinda confusing. Was hoping there was a central location.

**Reply by GlennBags** [reply]
Think I found it: https://github-wiki-see.page/m/squizlabs/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties

**Reply by GlennBags** [reply]
After testing, 
```
 <rule ref="PSR12">
     <exclude name="PEAR.ControlStructures.MultiLineCondition"/>
 </rule>
```
did not stop it from creating a new line after the opening `(`. 

However, the `PSR12.ControlStructures.ControlStructureSpacing.FirstExpressionLine` exclude did keep it from adjusting the line at all. We'll probably end up using the PSR-12 to be consistent, but it's still good to know where and how to apply these exceptions.

Thanks again for the help!

**Reply by jrfnl** [reply]
@GlennBags You seem to be mixing some things up.

>  is there a manual/listing where all of these sniffs are listed

No, there isn't. You found the list of all sniff with customizable properties, but that is not a complete list of sniffs, only a list of sniffs with properties you can change.

You can see which sniffs are included in each ruleset by using the `-e` (explain) option on the command-line, i.e. `phpcs --standard=PSR12 -e` will show you all sniffs which are included in the PSR12 standard.

> After testing,
> ```xml
>  <rule ref="PSR12">
>      <exclude name="PEAR.ControlStructures.MultiLineCondition"/>
>  </rule>
> ```
> did not stop it from creating a new line after the opening (.

It wouldn't as the PSR12 ruleset has its own sniff for control structures, so doesn't include the PEAR one.

What I meant by suggesting  that sniff, is to for you to replace the PSR12 one with the PEAR one:
```xml
<rule ref="PSR12">
    <!-- Remove the PSR12 control structure sniff. -->
    <exclude name="PSR12.ControlStructures.ControlStructureSpacing"/>
</rule>
<!-- Include the PEAR one instead. -->
<rule ref"PEAR.ControlStructures.MultiLineCondition"/>
```

**Reply by GlennBags** [reply]
Was going to say "I'm getting an error that `-e` is not valid.."; then I realized that you have it on `phpcs`, not `phpcbf`, which is what I'm using. I'm using the cbf since phpcs said it only "detects". I'm currently working on an old code base and trying to bring everything up to PSR-12. Are the sniffs that I'm getting out of phpcs the same as those used within phpcbf?

**Comment by jrfnl** [comment]
@GlennBags Yes, the sniffs are the same, `phpcs`, as you say, detects and reports on issues, `phpcbf` is a different way of invoking the same sniffs which will let the fixers included in the sniffs do their work.
